File: packages/react-native-nitro-device-integrity/android/src/main/java/com/margelo/nitro/nitrodeviceintegrity/DeviceIntegrity.kt
/**
 * DeviceIntegrity.kt - Android implementation (Play Integrity API)
 *
 * Implements the DeviceIntegrity HybridObject. Issues Play Integrity tokens
 * (Standard and Classic). This is a token-issuing client only: the returned
 * tokens are opaque, encrypted JWTs that MUST be verified by the developer's
 * server (Google `:decodeIntegrityToken` or self-managed keys).
 *
 * @author HyunWoo Lee
 */
package com.margelo.nitro.nitrodeviceintegrity

import android.content.Context
import com.facebook.proguard.annotations.DoNotStrip
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.tasks.Task
import com.google.android.play.core.integrity.IntegrityManagerFactory
import com.google.android.play.core.integrity.IntegrityServiceException
import com.google.android.play.core.integrity.IntegrityTokenRequest
import com.google.android.play.core.integrity.StandardIntegrityException
import com.google.android.play.core.integrity.StandardIntegrityManager.PrepareIntegrityTokenRequest
import com.google.android.play.core.integrity.StandardIntegrityManager.StandardIntegrityToken
import com.google.android.play.core.integrity.StandardIntegrityManager.StandardIntegrityTokenProvider
import com.google.android.play.core.integrity.StandardIntegrityManager.StandardIntegrityTokenRequest
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Android implementation of DeviceIntegrity backed by the Play Integrity API.
 */
@DoNotStrip
class DeviceIntegrity : HybridDeviceIntegritySpec() {
    /** React application context. */
    private val context: Context
        get() =
            NitroModules.applicationContext
                ?: throw IllegalStateException("React context not available")

    /** Cached Standard token provider, prepared once and reused. */
    @Volatile
    private var standardProvider: StandardIntegrityTokenProvider? = null

    /** Cloud project number used for the last successful prepare (for auto re-prepare). */
    @Volatile
    private var preparedCloudProjectNumber: Long? = null

    /** Serializes prepare/request so a provider is prepared at most once at a time. */
    private val providerMutex = Mutex()

    // MARK: - Availability

    /**
     * Best-effort availability hint: whether Google Play Services is present.
     * Does not guarantee the Play Integrity API is onboarded for this app.
     */
    override val isSupported: Boolean
        get() =
            runCatching {
                GoogleApiAvailability.getInstance()
                    .isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS
            }.getOrDefault(false)

    override val providerType: IntegrityProviderType
        get() =
            if (isSupported) {
                IntegrityProviderType.PLAYINTEGRITY
            } else {
                IntegrityProviderType.UNSUPPORTED
            }

    // MARK: - Play Integrity (Standard)

    override fun prepareStandardProvider(cloudProjectNumber: String): Promise<Unit> {
        return Promise.async {
            val number = parseCloudProjectNumber(cloudProjectNumber)
            providerMutex.withLock {
                prepareLocked(number)
            }
        }
    }

    override fun requestIntegrityToken(requestHash: String): Promise<String> {
        return Promise.async {
            try {
                requestStandardToken(requestHash)
            } catch (e: StandardIntegrityException) {
                // Provider expired -> re-prepare once and retry.
                if (e.errorCode == EXPIRED_PROVIDER_ERROR_CODE) {
                    val expiredProvider = standardProvider
                    val number =
                        preparedCloudProjectNumber
                            ?: throw integrityError(
                                "PROVIDER_NOT_PREPARED",
                                "Call prepareStandardProvider() before requestIntegrityToken().",
                                e,
                            )
                    providerMutex.withLock {
                        // Skip re-prepare if another coroutine already refreshed it.
                        if (standardProvider === expiredProvider) {
                            prepareLocked(number)
                        }
                    }
                    try {
                        requestStandardToken(requestHash)
                    } catch (retry: StandardIntegrityException) {
                        throw mapStandardException(retry)
                    }
                } else {
                    throw mapStandardException(e)
                }
            }
        }
    }

    /** Prepares the Standard provider. Caller must hold [providerMutex]. */
    private suspend fun prepareLocked(cloudProjectNumber: Long) {
        val manager = IntegrityManagerFactory.createStandard(context)
        val request =
            PrepareIntegrityTokenRequest.builder()
                .setCloudProjectNumber(cloudProjectNumber)
                .build()
        standardProvider =
            try {
                manager.prepareIntegrityToken(request).await()
            } catch (e: StandardIntegrityException) {
                throw mapStandardException(e)
            }
        preparedCloudProjectNumber = cloudProjectNumber
    }

    private suspend fun requestStandardToken(requestHash: String): String {
        val provider =
            standardProvider
                ?: throw integrityError(
                    "PROVIDER_NOT_PREPARED",
                    "Call prepareStandardProvider() before requestIntegrityToken().",
                )
        val request =
            StandardIntegrityTokenRequest.builder()
                .setRequestHash(requestHash)
                .build()
        val token: StandardIntegrityToken = provider.request(request).await()
        return token.token()
    }

    // MARK: - Play Integrity (Classic)

    override fun requestClassicIntegrityToken(
        nonce: String,
        cloudProjectNumber: String,
    ): Promise<String> {
        return Promise.async {
            val number = parseCloudProjectNumber(cloudProjectNumber)
            val manager = IntegrityManagerFactory.create(context)
            val request =
                IntegrityTokenRequest.builder()
                    .setNonce(nonce)
                    .setCloudProjectNumber(number)
                    .build()
            try {
                manager.requestIntegrityToken(request).await().token()
            } catch (e: IntegrityServiceException) {
                throw integrityError(
                    "CLASSIC_INTEGRITY_ERROR_${e.errorCode}",
                    e.message ?: "Classic integrity request failed",
                    e,
                )
            }
        }
    }

    // MARK: - iOS-only methods (rejected on Android)

    override fun generateKey(): Promise<String> = rejectIosOnly("generateKey")

    override fun attestKey(
        keyId: String,
        clientDataHash: String,
    ): Promise<String> = rejectIosOnly("attestKey")

    override fun generateAssertion(
        keyId: String,
        clientDataHash: String,
    ): Promise<String> = rejectIosOnly("generateAssertion")

    override fun getDeviceCheckToken(): Promise<String> = rejectIosOnly("getDeviceCheckToken")

    // MARK: - Helpers

    private fun <T> rejectIosOnly(method: String): Promise<T> =
        Promise.rejected(
            integrityError(
                "UNSUPPORTED_PLATFORM",
                "$method is only available on iOS (App Attest / DeviceCheck).",
            ),
        )

    private fun parseCloudProjectNumber(value: String): Long =
        value.trim().toLongOrNull()?.takeIf { it > 0 }
            ?: throw integrityError(
                "CLOUD_PROJECT_NUMBER_IS_INVALID",
                "cloudProjectNumber must be a positive integer string, got: $value",
            )

    private fun mapStandardException(e: StandardIntegrityException): Throwable =
        integrityError("STANDARD_INTEGRITY_ERROR_${e.errorCode}", e.message ?: "Standard integrity request failed", e)

    private fun integrityError(
        code: String,
        message: String,
        cause: Throwable? = null,
    ): Throwable = RuntimeException("$code: $message", cause)

    /** Bridges a Play Services [Task] into a coroutine. */
    private suspend fun <T> Task<T>.await(): T =
        suspendCancellableCoroutine { continuation ->
            addOnSuccessListener { result -> continuation.resume(result) }
            addOnFailureListener { error -> continuation.resumeWithException(error) }
            addOnCanceledListener { continuation.cancel() }
        }

    companion object {
        /** StandardIntegrityErrorCode.INTEGRITY_TOKEN_PROVIDER_INVALID (-19). */
        private const val EXPIRED_PROVIDER_ERROR_CODE = -19
    }
}

File: packages/react-native-nitro-device-integrity/ios/DeviceIntegrity.swift
/**
 * DeviceIntegrity.swift
 * React Native Nitro Device Integrity - iOS Implementation
 *
 * Implements the DeviceIntegrity HybridObject. Issues App Attest
 * (DCAppAttestService) attestations/assertions and DeviceCheck (DCDevice)
 * tokens. This is a token-issuing client only: every returned value is an
 * opaque blob (base64) that MUST be validated by the developer's server
 * against Apple's root CA / DeviceCheck API.
 *
 * @author HyunWoo Lee
 */

import Foundation
import NitroModules
import DeviceCheck

/// iOS implementation of DeviceIntegrity backed by App Attest + DeviceCheck.
class DeviceIntegrity: HybridDeviceIntegritySpec {

  // MARK: - Availability

  var isSupported: Bool {
    return DCAppAttestService.shared.isSupported
  }

  var providerType: IntegrityProviderType {
    return DCAppAttestService.shared.isSupported ? .appattest : .unsupported
  }

  // MARK: - App Attest

  func generateKey() throws -> Promise<String> {
    return Promise.async {
      guard DCAppAttestService.shared.isSupported else {
        throw self.unsupportedError("App Attest")
      }
      return try await withCheckedThrowingContinuation { continuation in
        DCAppAttestService.shared.generateKey { keyId, error in
          if let keyId = keyId {
            continuation.resume(returning: keyId)
          } else {
            continuation.resume(throwing: self.mapError(error, fallback: "Failed to generate App Attest key"))
          }
        }
      }
    }
  }

  func attestKey(keyId: String, clientDataHash: String) throws -> Promise<String> {
    return Promise.async {
      guard DCAppAttestService.shared.isSupported else {
        throw self.unsupportedError("App Attest")
      }
      let hash = try self.decodeBase64(clientDataHash, field: "clientDataHash")
      return try await withCheckedThrowingContinuation { continuation in
        DCAppAttestService.shared.attestKey(keyId, clientDataHash: hash) { attestation, error in
          if let attestation = attestation {
            continuation.resume(returning: attestation.base64EncodedString())
          } else {
            continuation.resume(throwing: self.mapError(error, fallback: "Failed to attest key"))
          }
        }
      }
    }
  }

  func generateAssertion(keyId: String, clientDataHash: String) throws -> Promise<String> {
    return Promise.async {
      guard DCAppAttestService.shared.isSupported else {
        throw self.unsupportedError("App Attest")
      }
      let hash = try self.decodeBase64(clientDataHash, field: "clientDataHash")
      return try await withCheckedThrowingContinuation { continuation in
        DCAppAttestService.shared.generateAssertion(keyId, clientDataHash: hash) { assertion, error in
          if let assertion = assertion {
            continuation.resume(returning: assertion.base64EncodedString())
          } else {
            continuation.resume(throwing: self.mapError(error, fallback: "Failed to generate assertion"))
          }
        }
      }
    }
  }

  // MARK: - DeviceCheck

  func getDeviceCheckToken() throws -> Promise<String> {
    return Promise.async {
      guard DCDevice.current.isSupported else {
        throw self.unsupportedError("DeviceCheck")
      }
      return try await withCheckedThrowingContinuation { continuation in
        DCDevice.current.generateToken { data, error in
          if let data = data {
            continuation.resume(returning: data.base64EncodedString())
          } else {
            continuation.resume(throwing: self.mapError(error, fallback: "Failed to generate DeviceCheck token"))
          }
        }
      }
    }
  }

  // MARK: - Android-only methods (rejected on iOS)

  func prepareStandardProvider(cloudProjectNumber: String) throws -> Promise<Void> {
    return Promise.rejected(withError: unsupportedError("Play Integrity (prepareStandardProvider)"))
  }

  func requestIntegrityToken(requestHash: String) throws -> Promise<String> {
    return Promise.rejected(withError: unsupportedError("Play Integrity (requestIntegrityToken)"))
  }

  func requestClassicIntegrityToken(nonce: String, cloudProjectNumber: String) throws -> Promise<String> {
    return Promise.rejected(withError: unsupportedError("Play Integrity (requestClassicIntegrityToken)"))
  }

  // MARK: - Helpers

  private func decodeBase64(_ value: String, field: String) throws -> Data {
    guard let data = Data(base64Encoded: value) else {
      throw NSError(
        domain: "DeviceIntegrity",
        code: 1001,
        userInfo: [NSLocalizedDescriptionKey: "INVALID_BASE64: \(field) is not valid base64"]
      )
    }
    return data
  }

  private func unsupportedError(_ feature: String) -> NSError {
    return NSError(
      domain: "DeviceIntegrity",
      code: 1002,
      userInfo: [NSLocalizedDescriptionKey: "UNSUPPORTED_PLATFORM: \(feature) is not available on this device/platform"]
    )
  }

  private func mapError(_ error: Error?, fallback: String) -> NSError {
    if let dcError = error as? DCError {
      let code: String
      switch dcError.code {
      case .invalidKey: code = "INVALID_KEY"
      case .invalidInput: code = "INVALID_INPUT"
      case .serverUnavailable: code = "SERVER_UNAVAILABLE"
      case .featureUnsupported: code = "FEATURE_UNSUPPORTED"
      case .unknownSystemFailure: code = "UNKNOWN_SYSTEM_FAILURE"
      @unknown default: code = "UNKNOWN"
      }
      return NSError(
        domain: "DeviceIntegrity",
        code: dcError.code.rawValue,
        userInfo: [
          NSLocalizedDescriptionKey: "\(code): \(dcError.localizedDescription)",
          NSUnderlyingErrorKey: dcError,
        ]
      )
    }
    if let error = error {
      return error as NSError
    }
    return NSError(
      domain: "DeviceIntegrity",
      code: -1,
      userInfo: [NSLocalizedDescriptionKey: fallback]
    )
  }
}
